taroscript/send.go

```diff
@@ -290,154 +164,181 @@ func IsValidInput(input *commitment.TaroCommitment,
+		locator := splitLocators[idx]
+		splitAsset, ok := splitCommitment.SplitAssets[*locator]
+		if !ok {
+			return fmt.Errorf("invalid split, asset for locator "+
+				"%v not found", locator)
+		}
 
-	// We'll now create a new copy of the old asset, swapping out the
-	// script key. We blank out the tweaked key information as this is now
-	// an external asset.
-	//
-	// TODO(roasbeef): make locators here, and make sure they exist like
-	// above
-	newAsset := updatedDelta.InputAssets[prevInput].Copy()
-	newAsset.ScriptKey.PubKey = &addr.ScriptKey
-	newAsset.ScriptKey.TweakedScriptKey = nil
-
-	// Record the PrevID of the input asset in a Witness for the new asset.
-	// This Witness still needs a valid signature for the new asset
-	// to be valid.
-	//
-	// TODO(roasbeef): when we fix #121, then this should also be a
-	// ZeroPrevID
-	newAsset.PrevWitnesses = []asset.Witness{
-		{
-			PrevID:          &prevInput,
-			TxWitness:       nil,
-			SplitCommitment: nil,
-		},
+		outputs[idx].Asset = &splitAsset.Asset
+		outputs[idx].Asset.ScriptKey = outputs[idx].ScriptKey
 	}
 
-	updatedDelta.NewAsset = *newAsset
-
-	return &updatedDelta
+	return nil
 }
 
-// CompleteAssetSpend updates the new Asset by creating a signature over the
-// asset transfer, verifying the transfer with the Taro VM, and attaching that
-// signature to the new Asset.
-func CompleteAssetSpend(internalKey btcec.PublicKey, delta SpendDelta,
-	signer Signer, validator TxValidator) (*SpendDelta, error) {
+// SignVirtualTransaction updates the new asset (the root asset located at the
+// change output in case of a non-interactive or partial amount send or the
+// full asset in case of an interactive full amount send) by creating a
+// signature over the asset transfer, verifying the transfer with the Taro VM,
+// and attaching that signature to the new Asset.
+func SignVirtualTransaction(input *taropsbt.VInput, outputs []*taropsbt.VOutput,
+	signer Signer, validator TxValidator) error {
 
-	updatedDelta := delta.Copy()
+	prevAssets := commitment.InputSet{
+		input.PrevID: input.Asset(),
+	}
+	newAsset := outputs[1].Asset
+	if input.IsSplit {
+		newAsset = outputs[0].Asset
+	}
 
 	// Create a Taro virtual transaction representing the asset transfer.
-	virtualTx, _, err := VirtualTx(
-		&updatedDelta.NewAsset, updatedDelta.InputAssets,
-	)
+	virtualTx, _, err := VirtualTx(newAsset, prevAssets)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	// For each input asset leaf, we need to produce a witness.
-	// Update the input of the virtual TX, generate a witness,
-	// and attach it to the copy of the new Asset.
-	validatedAsset := updatedDelta.NewAsset.Copy()
-	prevWitnessCount := len(updatedDelta.NewAsset.PrevWitnesses)
-
+	// For each input asset leaf, we need to produce a witness. Update the
+	// input of the virtual TX, generate a witness, and attach it to the
+	// copy of the new Asset.
+	//
+	// TODO(guggero): I think this is wrong... We shouldn't look at
```

Comment:
++ This needs a closer look.

Reply:
Copy pasting my comment from the other PR here: 

Yeah so we need to sign the new asset (which also anchors the split), along with the other leaves that occupied that same prev anchor output.

Here we're instead, re-signing all the old inputs for a merge (iiuc) which isn't correct.

In addition to just the Vinput, we also need the TaroCommitment for that, as for now we need to re-sign all the other inputs. FWIW, we only support spending a single input for now, so things are restricted to just what was in that sole tree.

---

Ah ok this was lifted from the old code. It's actually partially untested, since we only support a single input. This would be for the "merge" case, where we need to sign more than one taro input.

Once we lift the merge restriction (ends up needing a diff proof file creation), then this is equivalent to signing multiple Bitcoin inputs when you need to consolidate change to make a spend.

Reply:
I added a TODO in the hopefully right place.